### PR TITLE
Issue #359 - calculate Y

### PR DESCRIPTION
**Pull request by emanjon**
First take on explaining how to calculate a Y value for old APIs not supporting the full {{SECG}} format.

### CODE REVIEW

**Comment by emanjon** [comment]
Given e.g. the Julia programming language the calculations aretrivial

```
p = 2^BigInt(256) - 2^BigInt(224) + 2^BigInt(192) + 2^BigInt(96) - 1
a = BigInt(-3)
b = BigInt(41058363725152142129326129780047268409114441015993725554835256314039467401291)

X = BigInt(115792089183396302095546807154740558443406795108653336398970697772788799766525)
alpha = (X^3 + a * X + b) % p 
beta = powermod(alpha, (p + 1) ÷ 4, p)
```

For more constrained platforms maybe we should have some link to a paper explaining how to do this type of calculations on constrained platforms.

https://csrc.nist.gov/csrc/media/events/workshop-on-elliptic-curve-cryptography-standards/documents/papers/session6-adalier-mehmet.pdf

https://www.researchgate.net/publication/347528633_Fast_Implementation_of_NIST_P-256_Elliptic_Curve_Cryptography_on_8-Bit_AVR_Processor

file:///Users/emanjon/Downloads/applsci-10-08816.pdf

**Comment by emanjon** [comment]
`If a y-coordinate is required then the value ~yp SHALL be set to zero
`
This leads to unnecesary extra code if a y value is calculated. I suggest that we reformulate this

**Comment by emanjon** [comment]
Erik Thormarker wrote me that RFC 6090 Appendix C is a good high level reference for this.

**Comment by gselander** [comment]
@malishav and others: is the new text sufficient to support implementers which has to use a library requiring the y-coordinate?

File: draft-ietf-lake-edhoc.md [hunk 1]
```diff
@@ -1747,12 +1747,26 @@ In EDHOC, compact representation is used for the ephemeral public keys (G_X and
...
-* The compact representation described above can in such a case be transformed into the SECG point compressed format by prepending it with the single byte 0x02 (i.e., M = 0x02 \|\| X).
+The following may be needed for validation or compatibility with APIs that do not support compact representation or do not support the full {{SECG}} format:
 
-Using compact representation has some security benefits. An implementation does not need to check that the point is not the point at infinity (the identity element). Similarly, as not even the sign of the y-coordinate is encoded, compact representation trivially avoids so called "benign malleability" attacks where an attacker changes the sign, see {{SECG}}.
+* If a compressed y-coordinate is required, then the value ~yp set to zero can be used. The compact representation described above can in such a case be transformed into the SECG point compressed format by prepending it with the single byte 0x02 (i.e., M = 0x02 \|\| X).
+* If a uncompressed y-coordinate is required, then a y-coordinate has to be calculated following Section 2.3.4 of {{SECG}} or Appendix C of {{RFC6090}}. Any of the square roots can be used.
```
Comment (malishav) [inline draft-ietf-lake-edhoc.md:1755]: Not clear at this point what "any of the square roots" refers to. Consider expanding to "Any of the square root methods from {{REFS}} can be used".

File: draft-ietf-lake-edhoc.md [hunk 2]
```diff
@@ -571,7 +571,7 @@ The Initiator needs to have a list of cipher suites it supports in order of pref
...
+The ephemeral public keys in EDHOC (G_X and G_Y) use compact representation of elliptic curve points, see {{comrep}}. In COSE compact representation is achieved by formatting the ECDH ephemeral public keys as COSE_Keys of type EC2 or OKP according to Sections 7.1 and 7.2 of {{RFC9053}}, but only including the 'x' parameter in G_X and G_Y. For Elliptic Curve Keys of type EC2, compact representation MAY be used also in the COSE_Key. COSE always use compact output for Elliptic Curve Keys of type EC2. If the COSE implementation requires a 'y' parameter, the value y = false or a calculated y-coordinate can be used, see {{comrep}}.
```
Comment (malishav) [inline draft-ietf-lake-edhoc.md:574]: s/COSE always use/COSE always uses

File: draft-ietf-lake-edhoc.md [hunk 3]
```diff
@@ -1747,12 +1747,26 @@ In EDHOC, compact representation is used for the ephemeral public keys (G_X and
...
+
+* p = 2^256 − 2^224 + 2^192 + 2^96 − 1
+* a = -3
+* b = 41058363725152142129326129780047268409114441015993725554835256314039467401291
+
+Given an example x
+
+* x = 115792089183396302095546807154740558443406795108653336398970697772788799766525
+
+we can calculate y = w = (x^3 + a * x + b)^((p + 1)/4) (mod p)
```
Comment (malishav) [inline draft-ietf-lake-edhoc.md:1767]: Note somewhere that (p+1)/4 (mod p) is actually the square root operation, as explained in RFC6090.

**Pull request merged**